Tests/BSONTests/BSONTests.swift

```diff
@@ -1 +1,143 @@
+        if FileManager.default.fileExists(atPath: "./Tests") {
+            return "./Tests/Specs"
+        }
+        // otherwise we're in Xcode, get the bundle's resource path
+        guard let path = Bundle(for: self).resourcePath else {
+            XCTFail("Missing resource path")
+            return ""
+        }
+        return path
+    }
+
+    struct TestError: Error {
+        let message: String
+    }
+
+    func fail(_ message: String = "This should not happen") -> Never {
+        XCTFail(message)
+        fatalError(message)
+    }
+}
+
+public struct TestError: LocalizedError {
+    public let message: String
+    public var errorDescription: String { self.message }
+
+    public init(message: String) {
+        self.message = message
+    }
+}
+
+/// Given a spec folder name (e.g. "crud") and optionally a subdirectory name for a folder (e.g. "read") retrieves an
+/// array of [(filename, file decoded to type T)].
+public func retrieveSpecTestFiles<T: Decodable>(
+    specName: String,
+    subdirectory: String? = nil,
+    asType _: T.Type
+) throws -> [(String, T)] {
+    var path = "\(BSONTestCase.specsPath)/\(specName)"
+    if let sd = subdirectory {
+        path += "/\(sd)"
+    }
+    return try FileManager.default
+        .contentsOfDirectory(atPath: path)
+        .filter { $0.hasSuffix(".json") }
+        .map { filename in
+            let url = URL(fileURLWithPath: "\(path)/\(filename)")
+            let data = try Data(contentsOf: url, options: .mappedIfSafe)
+            let jsonResult = try JSONDecoder().decode(T.self, from: data)
+            return (filename, jsonResult)
+        }
+}
+
+public extension BSONDocument {
+    func toByteString() -> String {
+        guard let bytes = self.buffer.getBytes(at: 0, length: self.buffer.readableBytes) else {
+            return ""
+        }
+        return BSONTests.toByteString(bytes)
+    }
+
+    func readAllBytes() -> [UInt8] {
+        guard let bytes = self.buffer.getBytes(at: 0, length: self.buffer.readableBytes) else {
+            return []
+        }
+        return bytes
+    }
+}
+
+public func makeByteString(from bytes: [UInt8]) -> String {
+    var string = ""
+    for byte in bytes {
+        if (33 < byte) && (byte < 126) {
+            string += String(UnicodeScalar(byte))
+        } else {
+            string += "\\x" + String(format: "%02X", byte)
+        }
+    }
+    return string + ", \(String(format: "0x%02X", bytes.count))"
+}
+
+extension BSONDocument: NMBCollection {}
+
+func toByteString(_ bytes: [UInt8]?) -> String {
+    guard let defbytes = bytes else {
+        return "none"
+    }
+    var string = ""
+    for byte in defbytes {
+        if (33 < byte) && (byte < 126) {
+            string += String(UnicodeScalar(byte))
+        } else {
+            string += "\\x" + String(format: "%02X", byte)
+        }
+    }
+    return string
+}
+
+public func toByteString(buffer: ByteBuffer) -> String {
```

Comment:
I think it would be good if we can make the per-type helpers for this consistent, e.g. maybe all of the per-type methods should be instance methods defined in extensions, like you did for array

Reply:
I cleaned these up and changed the functionality a bit, right now there's no usage cus we didn't need the additional int tests I wrote but they come in handy while developing a BSON type for readable binary. I can remember (or make a ticket) to remove them when all is said and done